src/common/gcode_input_filter.hpp

```diff
@@ -0,0 +1,177 @@
+    ///          EFill::END - end of input reached
+    template <typename FREAD, typename FEOF>
+    inline EFill Fill(FREAD fread, FEOF feof) {
+        if (feof()) {
+            return END;
+        }
+        // advance "file position" by the last size of buffer read
+        startInputOffset += rdbufSize;
+        return fread(rdbuf, sizeof(rdbuf), &rdbufSize) ? OK : ERROR;
+    }
+
+    // STL-like interface
+    inline const char *begin() const { return rdbuf; }
+    inline const char *end() const { return begin() + size(); }
+    inline uint32_t size() const { return rdbufSize; }
+    inline void reset(uint32_t fileRestartOffset = 0) {
+        rdbufSize = 0;
+        r = rdbuf;
+        startInputOffset = fileRestartOffset;
+    }
+    inline uint32_t startOfs() const { return startInputOffset; }
+};
+
+/// GCodeInputFilter filters an input buffer of raw G-code text by stripping all comments
+/// @returns processed bytes
+///
+/// Sadly, the operation is not zero-copy since the G-code line must be in one continuous block to be handed over to Marlin
+/// thus crossing borders of the rdbuf is not easy to do now.
+///
+/// Memory requirements:
+/// GCODEBUFFSIZE + some state vars
+template <uint32_t GCODEBUFFSIZE = 97>
+class GCodeInputFilter {
+    char *g;
+    const char *r;
+    enum States { COMMENT,
+        COPY,
+        ONHOLD };
+    States state = COPY;
+    char gcbuf[GCODEBUFFSIZE];
+
+public:
+    GCodeInputFilter() {
+        reset();
+    }
+
+    /// resets the filter to initial state
+    inline void reset() {
+        g = gcbuf;
+        r = nullptr;
+        state = COPY;
+    }
+
+    /// Perform filtering of the input data
+    ///
+    /// RDBUF: basically anything that provides the following methods:
+    /// const char *begin()const
+    /// const char *end()const
+    /// uint32_t size()const
+    /// and a few others
+    ///
+    /// ENQUEUE_GCODE:
+    /// void enqueue_gcode(const char *gcodebuff, uint32_t new_file_offset)
+    /// shall take the content of gcodebuff (zero terminated) and push it into Marlin (or any other processing engine)
+    /// new_file_offset represents the offset in the input file/stream AFTER the gcode is pushed into Marlin
+    /// - that's because we want to track which commands were successfully pushed into Marlin and processed
+    ///   to know exactly where to continue in case of an input file/stream removal/error
+    ///
+    /// @returns -1 in case of being "on hold" - i.e. was unable to enqueue the G-code into Marlin (can be called again to retry).
+    ///          Otherwise it returns a positive number of input bytes processed.
+    ///          Please note - if the number != rdbuf.size(), the filter has reached a line which couldn't fit into Marlin's internal line buffer
+    ///             which implies the line would have been truncated and could possibly damage the machine being misinterpreted!
+    template <typename RDBUF, typename ENQUEUE_GCODE>
```

Comment:
Please try to cut the logic from the RDBuff and implement it into the filter. Therefore, we will be able to use any std container as a templated parameter. It radically increases the compatibility, usability and extendibility of the filter.